phoenix-core/src/main/java/org/apache/phoenix/iterate/SpoolingByteBufferSegmentQueue.java

```diff
@@ -0,0 +1,357 @@
+            if(!spoolTo.isInMemory()){
+                flushToDisk();
+            }
+            size++;
+        }
+
+
+        return result;
+    }
+
+    protected abstract boolean writeRecord(T t, OutputStream outputStream);
+
+    private void flushToMemory(){
+        byte[] data = spoolTo.getData();
+        chunk.resize(data.length);
+        spoolFrom = createInMemoryResultQueue(data, chunk);
+        GLOBAL_MEMORY_CHUNK_BYTES.update(data.length);
+        flushed = true;
+    }
+
+
+    private void flushToDisk(){
+        long sizeOfSpoolFile = spoolTo.getFile().length();
+        GLOBAL_SPOOL_FILE_SIZE.update(sizeOfSpoolFile);
+        GLOBAL_SPOOL_FILE_COUNTER.increment();
+        spoolFrom = createOnDiskResultQueue(spoolTo.getFile());
+        if (spoolTo.getFile() != null) {
+            spoolTo.getFile().deleteOnExit();
+        }
+        inMemByteSize = 0;
+        flushed = true;
+    }
+
+
+    public boolean isFlushed(){
+        return flushed;
+    }
+
+    public T peek() {
+        if(!flushed){
+            flushToMemory();
+        }
+        return spoolFrom.peek();
+    }
+
+    @Override
+    public T poll() {
+        if(!flushed){
+            flushToMemory();
+        }
+        return spoolFrom.poll();
+    }
+
+    public void close() throws IOException {
+        if(spoolFrom != null){
+            spoolFrom.close();
+        }
+    }
+
+    @Override
+    public Iterator<T> iterator() {
+        if(!flushed){
+            flushToMemory();
+        }
+        return spoolFrom.iterator();
+    }
+
+    @Override
+    public int size() {
+        return size ;
+    }
+
+    public long getInMemByteSize(){
+        return inMemByteSize;
+    };
+
+    private static abstract class ResultQueue<T> extends  AbstractQueue<T> implements  Closeable{}
+
+    protected static abstract class InMemoryResultQueue<T> extends ResultQueue<T> {
+        private final MemoryChunk memoryChunk;
+        protected final byte[] bytes;
+        private T next;
+        private AtomicInteger offset = new AtomicInteger(0);
+
+        protected InMemoryResultQueue(byte[] bytes, MemoryChunk memoryChunk) {
+            this.bytes = bytes;
+            this.memoryChunk = memoryChunk;
+            advance(offset);
+        }
+
+        protected abstract T advance(AtomicInteger offset);
+
+        @Override
+        public boolean offer(T t) {
+            return false;
+        }
+
+        @Override
+        public T peek(){
+            return next;
+        }
+
+        @Override
+        public T poll() {
+            T current = next;
+            next = advance(offset);
+            return current;
+        }
+
+
+        public void close() {
+            memoryChunk.close();
+        }
+
+
+        @Override
+        public Iterator<T> iterator() {
+            return new Iterator<T>(){
+                AtomicInteger iteratorOffset = new AtomicInteger(offset.get());
+                private T next = advance(iteratorOffset);
```

Comment:
One reminder: I said in my previous review that we'd better avoid discarding a old XXXSegmentQueue and starting a new one all the time when we switch between these segment queues. So we should ultimately remove this offset thing.